server/api_server/tools_routes.py

```diff
@@ -0,0 +1,686 @@
+    if not check_auth():
+        return jsonify({"error": "Unauthorized"}), 401
+
+    conn = get_temp_db_connection()
+    conn.row_factory = sqlite3.Row
+    cur = conn.cursor()
+
+    try:
+        cur.execute("SELECT devName, devMac, devLastIP as devIP, devVendor, devFirstConnection, devLastConnection FROM Devices ORDER BY devFirstConnection DESC")
+        rows = cur.fetchall()
+        devices = [dict(row) for row in rows]
+        return jsonify(devices)
+    except Exception as e:
+        return jsonify({"error": str(e)}), 500
+    finally:
+        conn.close()
+
+
+@tools_bp.route('/get_device_info', methods=['POST'])
+def get_device_info():
+    """Get detailed info for a specific device."""
+    if not check_auth():
+        return jsonify({"error": "Unauthorized"}), 401
+
+    data = request.get_json()
+    if not data or 'query' not in data:
+        return jsonify({"error": "Missing 'query' parameter"}), 400
+
+    query = data['query']
+
+    conn = get_temp_db_connection()
+    conn.row_factory = sqlite3.Row
+    cur = conn.cursor()
+
+    try:
+        # Search by MAC, Name, or partial IP
+        sql = "SELECT * FROM Devices WHERE devMac LIKE ? OR devName LIKE ? OR devLastIP LIKE ?"
+        cur.execute(sql, (f"%{query}%", f"%{query}%", f"%{query}%"))
+        rows = cur.fetchall()
+
+        if not rows:
+            return jsonify({"message": "No devices found"}), 404
+
+        devices = [dict(row) for row in rows]
+        return jsonify(devices)
+    except Exception as e:
+        return jsonify({"error": str(e)}), 500
+    finally:
+        conn.close()
+
+
+@tools_bp.route('/get_latest_device', methods=['POST'])
+def get_latest_device():
+    """Get full details of the most recently discovered device."""
+    if not check_auth():
+        return jsonify({"error": "Unauthorized"}), 401
+
+    conn = get_temp_db_connection()
+    conn.row_factory = sqlite3.Row
+    cur = conn.cursor()
+
+    try:
+        # Get the device with the most recent devFirstConnection
+        cur.execute("SELECT * FROM Devices ORDER BY devFirstConnection DESC LIMIT 1")
+        row = cur.fetchone()
+
+        if not row:
+            return jsonify({"message": "No devices found"}), 404
+
+        # Return as a list to be consistent with other endpoints
+        return jsonify([dict(row)])
+    except Exception as e:
+        return jsonify({"error": str(e)}), 500
+    finally:
+        conn.close()
+
+
+@tools_bp.route('/get_open_ports', methods=['POST'])
+def get_open_ports():
+    """
+    Specific query for the port-scan results of a target.
+    Arguments: target (IP or MAC)
+    """
+    if not check_auth():
+        return jsonify({"error": "Unauthorized"}), 401
+
+    data = request.get_json()
+    target = data.get('target')
+
+    if not target:
+        return jsonify({"error": "Target is required"}), 400
+
+    # If MAC is provided, try to resolve to IP
+    if re.match(r"^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})$", target):
```

Comment:
there is already an plugin_helper.is_mac method - let's try to follow the DRY principle